### ISSUE TITLE
linter: typescript/no-misused-new false positive

### ISSUE THREAD

**Issue by brianjhanson** [oxc-project/oxc#21990]
I'm not a total typescript guru, so this could definitely be a "me" problem but based on the [docs for the `typescript/no-misused-new` rule](https://oxc.rs/docs/guide/usage/linter/rules/typescript/no-misused-new.html) I would expect this example

```ts
interface TestInterface {
  new (config: Record<string, any>): TestInterface;
}

interface SomeObject {
  Test: TestInterface
}

class TestClass {
  constructor(config: Record<string, any>) {
    console.log({config});
  }
}

const someObject: SomeObject = {
  Test: TestClass
}

new someObject.Test({config: 'value'});
```

Specifically because of the second example of correct code:

```ts
interface I {
  new (): C;
}
```

Instead, I'm seeing an Oxc warning

> Oxc Warning: Interfaces cannot be constructed, only classes.

---

[Playground link](https://playground.oxc.rs/?options=%7B%22run%22%3A%7B%22lint%22%3Atrue%2C%22formatter%22%3Afalse%2C%22transform%22%3Afalse%2C%22isolatedDeclarations%22%3Afalse%2C%22whitespace%22%3Afalse%2C%22mangle%22%3Afalse%2C%22compress%22%3Afalse%2C%22scope%22%3Atrue%2C%22symbol%22%3Atrue%2C%22cfg%22%3Afalse%7D%2C%22parser%22%3A%7B%22extension%22%3A%22ts%22%2C%22allowReturnOutsideFunction%22%3Atrue%2C%22preserveParens%22%3Atrue%2C%22allowV8Intrinsics%22%3Atrue%2C%22semanticErrors%22%3Atrue%7D%2C%22linter%22%3A%7B%7D%2C%22formatter%22%3A%7B%22useTabs%22%3Afalse%2C%22tabWidth%22%3A2%2C%22endOfLine%22%3A%22lf%22%2C%22printWidth%22%3A80%2C%22singleQuote%22%3Afalse%2C%22jsxSingleQuote%22%3Afalse%2C%22quoteProps%22%3A%22as-needed%22%2C%22trailingComma%22%3A%22all%22%2C%22semi%22%3Atrue%2C%22arrowParens%22%3A%22always%22%2C%22bracketSpacing%22%3Atrue%2C%22bracketSameLine%22%3Afalse%2C%22objectWrap%22%3A%22preserve%22%2C%22singleAttributePerLine%22%3Afalse%7D%2C%22transformer%22%3A%7B%22target%22%3A%22es2015%22%2C%22useDefineForClassFields%22%3Atrue%2C%22experimentalDecorators%22%3Atrue%2C%22emitDecoratorMetadata%22%3Atrue%2C%22optimizeEnums%22%3Atrue%2C%22optimizeConstEnums%22%3Atrue%7D%2C%22isolatedDeclarations%22%3A%7B%22stripInternal%22%3Afalse%7D%2C%22codegen%22%3A%7B%22normal%22%3Atrue%2C%22jsdoc%22%3Atrue%2C%22annotation%22%3Atrue%2C%22legal%22%3Atrue%7D%2C%22compress%22%3A%7B%7D%2C%22mangle%22%3A%7B%22topLevel%22%3Atrue%2C%22keepNames%22%3Afalse%7D%2C%22controlFlow%22%3A%7B%22verbose%22%3Afalse%7D%2C%22inject%22%3A%7B%22inject%22%3A%7B%7D%7D%2C%22define%22%3A%7B%22define%22%3A%7B%7D%7D%7D&code=%0A%0Ainterface+TestInterface+%7B%0A++new+%28config%3A+Record%3Cstring%2C+any%3E%29%3A+TestInterface%0A%7D%0A%0Ainterface+SomeObject+%7B%0A++Test%3A+TestInterface%0A%7D%0A%0Aclass+TestClass+%7B%0A++constructor%28config%3A+Record%3Cstring%2C+any%3E%29+%7B%0A++++console.log%28%7Bconfig%7D%29%3B%0A++%7D%0A%7D%0A%0Aconst+someObject%3A+SomeObject+%3D+%7B%0A++Test%3A+TestClass%0A%7D%0A%0Anew+someObject.Test%28%7Bconfig%3A+%27value%27%7D%29%3B)

**Comment by camc314** [comment]
Hi! thanks for the feedback.

This is expected behavior rather than a false positive.

`TestInterface` is describing the constructor/static side, but its construct signature returns `TestInterface` itself:

```ts
interface TestInterface {
  new (config: Record<string, any>): TestInterface;
}
```

That is what `typescript/no-misused-new` reports. The construct signature should return the instance type created by the constructor, for example:

```ts
interface TestInstance {}

interface TestConstructor {
  new (config: Record<string, any>): TestInstance;
}
```

This behaviour matches the upstream [typescript-eslint/no-misused-new](https://typescript-eslint.io/rules/no-misused-new/) rule; their rule also reports the self-returning interface construct signature.

That said, the oxc docs were ambiguous because the invalid `new (): I` example was bundled with another invalid `constructor()` member. I've opened #22016 to update the docs to better describe the behaviour.

Thanks!

**Comment by brianjhanson** [comment]
Thanks for the explanation and docs update @camc314!
